glslc/src/dependency_info.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015 The Shaderc Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GLSLC_DEPENDENCY_INFO_H
+#define GLSLC_DEPENDENCY_INFO_H
+
+#include <set>
+#include <string>
+
+namespace glslc {
+class DependencyInfoDumpingHandler {
+ public:
+  // Sets -MT argument.
+  void SetDepTargetFileLabel(const std::string& target_label) {
+    dep_target_label_ = std::string(target_label);
+  }
+
+  // Sets -MF argument.
+  void SetDepFileName(const std::string& dep_file_name) {
+    dep_file_name_ = dep_file_name;
+  };
+
+  // Collects the dependency info and dump it to dependency info file.
+  bool DumpDependencyInfo(std::string target_file, std::string source_file_name,
+                          const std::set<std::string>& trace);
+
+  // -MM or -M will prevent emitting the output of compilation.
+  bool dump_dependency_info_only_ = false;
```

Comment:
Do you really need two booleans?
If dependency info is being generated, either
 (a) it's the main output of the compiler and we also set preprocess-only (the -M and -MM cases)
or  (b) it's written out separately, and the other parts of the compiler behave normally (the -MD and -MMD cases).

I have to see the higher level algorithm to see if it makes sense.